sgd.lua

```diff
@@ -53,6 +55,13 @@ function optim.sgd(opfunc, x, config, state)
       dfdx:add(state.decayParameters)
    end
 
+   -- (2.1) L1 regularization
+   if L1coeff > 0 then
+       dfdx:add(L1coeff, torch.sign(x))
```

Comment:
torch.sign(x) allocates a new result tensor at every call. This is bad practice, you ideally want to reuse the same buffer every time (see line 51 to 55 in your version).